examples/userapi/subdomains.ipynb

```diff
@@ -0,0 +1,966 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class FullDomain(SubDomain):\n",
+    "    name = 'mydomain'\n",
+    "    def define(self, dimensions):\n",
+    "        x, y, z = dimensions\n",
+    "        return {x: x, y: y, z: z}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the above snippet, we have defined a class `FullDomain` based on `SubDomain` with the method `define` overridden. This method takes as input a set of Dimensions and produces a mapper from which we can create a subdomain. It is through utilizing this mapper that various types of subdomains can be created. The mapper `{x: x, y: y, z: z}` is simply saying that we wish for the three dimensions of our subdomain be exactly the input dimensions `x, y, z`. Note that for a 2D domain we would need to remove one of the dimensions, i.e. `{x: x, y: y}`, or more generally for an N-dimensional domain one could write `{d: d for d in dimensions}`.\n",
+    "\n",
+    "Now, lets us compose a subdomain from our class `FullDomain`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "full_domain = FullDomain()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It is important to note that whilst the `full_domain` object has now been initialised, it has not yet been finalised, i.e. it is not currently associated with any grid. To attach this subdomain to a grid we must pass it to the grid during its creation (the sudomain will then be finalised during creation of the grid):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "my_grid = Grid(shape = (10, 10, 10), subdomains = (full_domain, ))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we look at the subdomains of `my_grid` we see that our new subdomain 'mydomain' is now present:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'domain': SubDomain domain[(x, y, z)],\n",
+       " 'interior': SubDomain interior[(xi, yi, zi)],\n",
+       " 'mydomain': SubDomain mydomain[(x, y, z)]}"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "my_grid.subdomains"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As expected (and intended), the dimensions of our newly defined subdomain match those of 'domain'.\n",
+    "\n",
+    "Now, let us create a subdomain consisting of all but the outer grid point in each dimension (similar to 'interior'):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Innerdomain(SubDomain):\n",
```

Comment:
nitpicking: `InnerDomain` for homogeneity

Reply:
Changed.